Visual preload tags

lemoswilson opened the pull request:
Adding support for responsive preload tags. Will update the linked `vue-visual` package to point to the main branch once https://github.com/BKWLD/vue-visual/pull/76 is merged, so that should be reviewed first.

weotch commented on mixins/visual/responsive-visual.vue:
```diff
@@ -33,6 +56,7 @@ export default
 		@isLanscapeMediaQuery?.removeListener @checkIsLandscape
 
 	computed:
+		isResponsiveImage: -> !!@landscape?.props?.image and !!@portrait?.props?.image and @preload
```
Is the `and @preload` required here?  I would think, based on the name, only the checks for landscape and portrait images would be required.

weotch commented on mixins/visual/responsive-visual.vue:
```diff
@@ -19,6 +19,29 @@ export default
 	}
 
+	head: ->
+		return {} unless @isResponsiveImage and @preload
```
It's ok to return void from this

```suggestion
		return unless @isResponsiveImage and @preload
```

weotch commented on mixins/visual/responsive-visual.vue:
```diff
@@ -19,6 +19,29 @@ export default
+				as: 'image'
+				imagesrcset: @makeSrcset @landscape.props.image
+				href: @landscape?.props?.image?.url
+				imagesizes: @sizes
```
Prevent an attribute value of `undefined` from being rendered when this is empty

```suggestion
				imagesizes: @sizes || ''
```

weotch commented on mixins/visual/responsive-visual.vue:
```diff
@@ -19,6 +19,29 @@ export default
+				as: 'image'
+				imagesrcset: @makeSrcset @portrait.props.image
+				href: @portrait?.props?.image?.url
+				imagesizes: @sizes,
```
Prevent an attribute value of `undefined` from being rendered when this is empty

```suggestion
				imagesizes: @sizes || ''
```

weotch reviewed (changes requested):
In addition to these little tweaks, I recommend updating `components/cloak-visual.coffee` with changes similar to [this](https://github.com/BKWLD/cloak-visual/blob/bcaeeb1cf40d6c4a1df0b92bc029018abd1a93ba/components/visual.coffee#L106-L108) 

```coffee
isCriticalImage = injections.blockIndex < 2
preload = props.preload ? isCriticalImage
lazyload = props.lazyload ? not isCriticalImage && not preload
```

So that:

- `preload` is automatically enabled for "above the fold" blocks
- `lazyload` is automatically disabled whenever `preload` is enabled

weotch commented:
I made another change to @cloak-app/visual that you might want for this PR: https://github.com/BKWLD/cloak-visual/pull/24.  This makes it so responsive visuals read the `blockIndex` provided value and auto-set preload state.

lemoswilson commented:
Thanks for reviewing the changes. Just pushed code addressing the review topics. Should've commited the suggestions using the GitHub web page UI, instead I went and made changes through the code and pushed as only one commit 🤦🏽‍♂️. But yeah, we should be good to merge it now.

weotch commented on components/cloak-visual.coffee:
```diff
@@ -104,7 +104,8 @@ export default
 		# kinda weird so it defaults to true when blockIndex is undefined
 		isCriticalImage = injections.blockIndex < 2
-		lazyload = props.lazyload ? not isCriticalImage
+		preload = props.preload ? isCriticalImage
```
You'll need to add a `preload` prop [like this](https://github.com/BKWLD/cloak-visual/blob/1e9ccf0ee0fa9db669b24d0ad07eba640a8d9e61/components/visual.coffee#L43-L45) too, otherwise `props.preload === false`.  We want it to `=== undefined` for use in this ` ? ` test.

lemoswilson replied:
`preload` is part of `vue-visual` [props](https://github.com/BKWLD/vue-visual/blob/master/concerns/supports-images.coffee#L10) since merging the support [preload PR](https://github.com/BKWLD/vue-visual/pull/76), and `cloak-visual` reduces  `vue-visual` mixin's props and [inherits those props](https://github.com/BKWLD/cloak/pull/63/files/cde632ab778c505bcac31a202a380b8b9800b656#diff-d8e7810cdaeec713ccc3cc13143336abb263431e9ca81325b93a4e33e6872fa1R9-R23), so `props.preload` should be `undefined` if no `preload` value is passed as prop.

weotch replied:
Well, the prob is that in `vue-visual`, it's a straight up Boolean prop without an explicit default, so it defaults to `false` when undefined.  So without overriding the prop to have a an _explicit_ default of `undefined`, `props.preload` will  `=== false` whenever it's not specified.

lemoswilson replied:
Yeah you are right, it does default to `false`. I believed that every defined prop without a set default would take `undefined` if no value was passed, but that's not the case with Boolean props. Is it the only type that behaves like this? Just pushed a change adding a `preload` that defaults to undefined.

weotch replied:
Yeah, I think it's just `Boolean` that has this kinda behavior.  Maybe `String` defaults to `""`?  Not 💯 about that.

weotch commented on mixins/visual/responsive-visual.vue:
```diff
@@ -110,13 +140,20 @@ export default
 				...@$props
 				image
 				video
+				preload
```
I think you should auto turn off lazyloading, too, when preloading, [like here](https://github.com/BKWLD/cloak-visual/blob/1e9ccf0ee0fa9db669b24d0ad07eba640a8d9e61/components/responsive.vue#L157-L160).

I added this stuff this morning in https://github.com/BKWLD/cloak-visual/pull/24, realizing that responsive visuals wouldn't auto set the preload attributes like we want.

lemoswilson replied:
just pushed a [change](https://github.com/BKWLD/cloak/pull/63/commits/cae51c314da13a782571649615f4e1734e672fab) to turn off lazyloading when preloading.
